File: PythonArchive/Shortest-Peak-Reversal-5m-3x-Margin--main/Unity App Project/short_trader_multi_filter/Assets/Scripts/Core/BuyOrderEngine.cs
using System;
using System.Collections.Generic;

namespace UnityApp.ShortTraderMultiFilter
{
    public class BuyOrderEngine
    {
        private readonly TraderConfig _config;

        public BuyOrderEngine(TraderConfig config)
        {
            _config = config;
        }

        public bool ShouldEnter(Dictionary<string, object> row, string? positionSide)
        {
            if (!string.IsNullOrEmpty(positionSide))
            {
                return false;
            }

            return row.TryGetValue("entry_signal", out var signalObj) && Convert.ToBoolean(signalObj)
                && (!row.TryGetValue("tradable", out var tradableObj) || Convert.ToBoolean(tradableObj));
        }

        public (PositionState? Position, string Status, double EntryFee, double TradeValue, double MarginUsed) OpenPosition(
            Candle candle,
            double availableUsdt)
        {
            var (entryPrice, status) = OrderUtils.SimulateOrderFill("long", candle.Close, _config);
            if (status == "rejected" || !entryPrice.HasValue)
            {
                return (null, status, 0, 0, 0);
            }

            if (availableUsdt <= 0 || _config.RiskFraction <= 0)
            {
                return (null, "insufficient_funds", 0, 0, 0);
            }

            var marginUsed = availableUsdt * _config.RiskFraction;
            var leverageUsed = OrderUtils.ResolveLeverage(marginUsed, _config.DesiredLeverage, _config);
            var tradeValue = marginUsed * leverageUsed;
            if (tradeValue < _config.MinNotional)
            {
                return (null, "min_notional_not_met", 0, 0, 0);
            }

            var qty = tradeValue / entryPrice.Value;
            var entryFee = OrderUtils.BybitFee(tradeValue, _config);
            var liqPrice = OrderUtils.CalcLiquidationPriceLong(entryPrice.Value, (int)leverageUsed);

            var position = new PositionState
            {
                Side = "long",
                EntryPrice = entryPrice.Value,
                LiquidationPrice = liqPrice,
                Quantity = qty,
                EntryTime = candle.Timestamp,
                EntryFee = entryFee,
                TradeValue = tradeValue,
                MarginUsed = marginUsed,
                Leverage = leverageUsed
            };
            return (position, status, entryFee, tradeValue, marginUsed);
        }
    }
}

File: PythonArchive/Shortest-Peak-Reversal-5m-3x-Margin--main/src/short_trader_multi_filter/buy_order_engine.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional, Tuple

import pandas as pd

from .config import TraderConfig
from .order_utils import PositionState, bybit_fee_fn, calc_liq_price_long, resolve_leverage, simulate_order_fill


@dataclass
class BuyOrderEngine:
    config: TraderConfig

    def should_enter(self, row: pd.Series, position_side: Optional[str]) -> bool:
        if position_side:
            return False
        return bool(row.get("entry_signal") and row.get("tradable", True))

    def open_position(
        self,
        row: pd.Series,
        available_usdt: float,
    ) -> Tuple[Optional[PositionState], str, float, float, float]:
        entry_price, status = simulate_order_fill("long", row["Close"], self.config)
        if status == "rejected" or entry_price is None:
            return None, status, 0.0, 0.0, 0.0

        if available_usdt <= 0:
            return None, "insufficient_funds", 0.0, 0.0, 0.0

        risk_fraction = min(max(self.config.risk_fraction, 0.0), 1.0)
        if risk_fraction == 0:
            return None, "insufficient_funds", 0.0, 0.0, 0.0

        margin_used = available_usdt * risk_fraction
        leverage_used = resolve_leverage(margin_used, self.config.desired_leverage, self.config)
        trade_value = margin_used * leverage_used
        if trade_value < self.config.min_notional:
            return None, "min_notional_not_met", 0.0, 0.0, 0.0
        qty = trade_value / entry_price
        entry_fee = bybit_fee_fn(trade_value, self.config)

        liq_price = calc_liq_price_long(entry_price, int(leverage_used))

        position = PositionState(
            side="long",
            entry_price=entry_price,
            liq_price=liq_price,
            qty=qty,
            entry_bar_time=row.name,
            entry_fee=entry_fee,
            trade_value=trade_value,
            margin_used=margin_used,
            leverage=leverage_used,
        )
        return position, status, entry_fee, trade_value, margin_used
